Support Cabal 3.12

It seems like uuagc needs to be adjusted for Cabal >= 3.12 due to the introduction of the `Suffix` newtype: https://github.com/haskell/cabal/commit/65905b980e5c79362cfe62416720aee3f78e5628


Currently compilation fails with

```
[4 of 5] Compiling Distribution.Simple.UUAGC.UUAGC ( src/Distribution/Simple/UUAGC/UUAGC.hs, dist/build/Distribution/Simple/UUAGC/UUAGC.o, dist/build/Distribution/Simple/UUAGC/UUAGC.dyn_o )
src/Distribution/Simple/UUAGC/UUAGC.hs:109:52: error: [8;;https://errors.haskell.org/messages/GHC-83865GHC-838658;;]
    • Couldn't match type ‘[Char]’ with ‘Suffix’
      Expected: Suffix
        Actual: String
    • In the expression: "ag"
      In the first argument of ‘(:)’, namely ‘("ag", ag)’
      In the expression: ("ag", ag) : ("lag", ag) : knownSuffixHandlers
    |
109 |   hooks = simpleUserHooks { hookedPreProcessors = ("ag", ag):("lag",ag):knownSuffixHandlers
    |                                                    ^^^^

src/Distribution/Simple/UUAGC/UUAGC.hs:109:63: error: [8;;https://errors.haskell.org/messages/GHC-83865GHC-838658;;]
    • Couldn't match type ‘[Char]’ with ‘Suffix’
      Expected: Suffix
        Actual: String
    • In the expression: "lag"
      In the first argument of ‘(:)’, namely ‘("lag", ag)’
      In the second argument of ‘(:)’, namely
        ‘("lag", ag) : knownSuffixHandlers’
    |
109 |   hooks = simpleUserHooks { hookedPreProcessors = ("ag", ag):("lag",ag):knownSuffixHandlers
    |                                                               ^^^^^
```

--- a/uuagc/trunk/cabal-plugin/src/Distribution/Simple/UUAGC/UUAGC.hs
+++ b/uuagc/trunk/cabal-plugin/src/Distribution/Simple/UUAGC/UUAGC.hs
@@ -1,4 +1,4 @@
-{-# LANGUAGE CPP #-}
+{-# LANGUAGE CPP, OverloadedStrings, DataKinds #-}
 module Distribution.Simple.UUAGC.UUAGC(uuagcUserHook,
                                        uuagcUserHook',
                                        uuagc,
@@ -56,7 +56,9 @@ import Data.List (nub,intersperse)
 import Data.Map (Map)
 import qualified Data.Map as Map
 
-#if MIN_VERSION_Cabal(3,6,0)
+#if MIN_VERSION_Cabal(3,14,0)
+import Distribution.Utils.Path (getSymbolicPath, Pkg, Source, SymbolicPath, FileOrDir (Dir), interpretSymbolicPathCWD)
+#elif MIN_VERSION_Cabal(3,6,0)
 import Distribution.Utils.Path (getSymbolicPath, PackageDir, SourceDir, SymbolicPath)
 #endif
 
@@ -192,6 +194,13 @@ getOptionsFromClass classes fOpt =
                                                    ++ show fClass
                                                    ++ " is not defined."
 
+#if MIN_VERSION_Cabal(3,14,0)
+buildDir' :: LocalBuildInfo -> FilePath
+buildDir' = interpretSymbolicPathCWD . buildDir
+#else
+buildDir' = buildDir
+#endif
+
 -- uuagcSDistHook :: ([String] -> FilePath -> IO (ExitCode, [FilePath]))
 --      -> PackageDescription
 --      -> Maybe LocalBuildInfo
@@ -202,7 +211,7 @@ getOptionsFromClass classes fOpt =
 --   {-
 --   case mbLbi of
 --     Nothing -> warn normal "sdist: the local buildinfo was not present. Skipping AG initialization. Dist may fail."
---     Just lbi -> let classesPath = buildDir lbi </> agClassesFile
+--     Just lbi -> let classesPath = buildDir' lbi </> agClassesFile
 --                 in commonHook uuagc classesPath pd lbi (sDistVerbosity df)
 --   originalSDistHook pd mbLbi uh df
 --   -}
@@ -216,7 +225,7 @@ uuagcBuildHook
      -> BuildFlags
      -> IO ()
 uuagcBuildHook uuagc pd lbi uh bf = do
-  let classesPath = buildDir lbi </> agClassesFile
+  let classesPath = buildDir' lbi </> agClassesFile
   commonHook uuagc classesPath pd lbi (buildVerbosity bf)
   originalBuildHook pd lbi uh bf
 
@@ -229,7 +238,7 @@ commonHook :: ([String] -> FilePath -> IO (ExitCode, [FilePath]))
 commonHook uuagc classesPath pd lbi fl = do
   let verbosity = fromFlagOrDefault normal fl
   info verbosity $ "commonHook: Assuming AG classesPath: " ++ classesPath
-  createDirectoryIfMissingVerbose verbosity True (buildDir lbi)
+  createDirectoryIfMissingVerbose verbosity True (buildDir' lbi)
   -- Read already existing options
   -- Map FilePath (Options, Maybe (FilePath,[String]))
   oldOptions <- readFileOptions classesPath
@@ -272,7 +281,7 @@ uuagc' uuagc build lbi _ =
      platformIndependent = True,
      runPreProcessor = mkSimplePreProcessor $ \ inFile outFile verbosity ->
                        do notice verbosity $ "[UUAGC] processing: " ++ inFile ++ " generating: " ++ outFile
-                          let classesPath = buildDir lbi </> agClassesFile
+                          let classesPath = buildDir' lbi </> agClassesFile
                           info verbosity $ "uuagc-preprocessor: Assuming AG classesPath: " ++ classesPath
                           fileOpts <- readFileOptions classesPath
                           opts <- case Map.lookup inFile fileOpts of
@@ -290,7 +299,10 @@ uuagc' uuagc build lbi _ =
 
 -- | In Cabal 3.6.0.0 (GHC 9.2) and up, 'BuildInfo' member 'hsSourceDirs' has type
 -- '[SymbolicPath PackageDir SourceDir]', but in versions before that, it is [FilePath].
-#if MIN_VERSION_Cabal(3,6,0)
+#if MIN_VERSION_Cabal(3,14,0)
+hsSourceDirsFilePaths :: [SymbolicPath Pkg (Dir Source)] -> [FilePath]
+hsSourceDirsFilePaths = map getSymbolicPath
+#elif MIN_VERSION_Cabal(3,6,0)
 hsSourceDirsFilePaths :: [SymbolicPath PackageDir SourceDir] -> [FilePath]
 hsSourceDirsFilePaths = map getSymbolicPath
 #else

--- a/uuagc/trunk/cabal-plugin/uuagc-cabal.cabal
+++ b/uuagc/trunk/cabal-plugin/uuagc-cabal.cabal
@@ -17,7 +17,7 @@ tested-with: GHC >= 6.12
 extra-source-files: README
 
 library
-   build-depends:    base >= 4, base < 5, Cabal >= 2, directory >= 1.0.1.1
+   build-depends:    base >= 4, base < 5, Cabal >= 2.0 && <3.15, directory >= 1.0.1.1
    build-depends:    process >= 1.0.1.3, containers >= 0.3, uulib >= 0.9.14, filepath >= 1.1.0.4, mtl >= 2.2.1
    hs-source-dirs:   src, src-options
    default-language: Haskell2010

--- a/uuagc/trunk/cabal.project
+++ b/uuagc/trunk/cabal.project
@@ -1,0 +1,1 @@
+packages: ., cabal-plugin

--- a/uuagc/trunk/uuagc.cabal
+++ b/uuagc/trunk/uuagc.cabal
@@ -33,7 +33,7 @@ flag with-loag
    manual: True
 
 custom-setup
-  setup-depends: base >= 4 && < 5, Cabal >= 1.24 && < 3.11, uuagc-cabal >= 1.0
+  setup-depends: base >= 4 && < 5, Cabal >= 2.0 && < 3.15, uuagc-cabal >= 1.0
 
 executable uuagc
    build-depends: uuagc-cabal >= 1.0.3.0

